go/billing/api.py

```diff
@@ -141,47 +142,42 @@ def render_POST(self, request):
-                 session_created):
+    def get_cost(self, account_number, tag_pool_name, provider,
+                 message_direction, session_created):
         """Return the message cost"""
         query = """
-            SELECT t.account_number, t.tag_pool_name, t.message_direction,
+            SELECT t.account_number, t.tag_pool_name,
+                   t.provider, t.message_direction,
                    t.message_cost, t.storage_cost, t.session_cost,
                    t.markup_percent
             FROM (SELECT a.account_number, t.name AS tag_pool_name,
-                         c.message_direction, c.message_cost, c.storage_cost,
-                         c.session_cost, c.markup_percent
+                         c.provider, c.message_direction,
+                         c.message_cost, c.storage_cost, c.session_cost,
+                         c.markup_percent
                   FROM billing_messagecost c
-                  INNER JOIN billing_tagpool t ON (c.tag_pool_id = t.id)
-                  INNER JOIN billing_account a ON (c.account_id = a.id)
-                  WHERE a.account_number = %(account_number)s
-                  AND t.name = %(tag_pool_name)s
-                  AND c.message_direction = %(message_direction)s
-                  UNION
-                  SELECT NULL AS account_number, t.name AS tag_pool_name,
-                         c.message_direction, c.message_cost, c.storage_cost,
-                         c.session_cost, c.markup_percent
-                  FROM billing_messagecost c
-                  INNER JOIN billing_tagpool t ON (c.tag_pool_id = t.id)
-                  WHERE c.account_id IS NULL
-                  AND t.name = %(tag_pool_name)s
-                  AND c.message_direction = %(message_direction)s
-                  UNION
-                  SELECT NULL AS account_number, NULL AS tag_pool_name,
-                         c.message_direction, c.message_cost, c.storage_cost,
-                         c.session_cost, c.markup_percent
-                  FROM billing_messagecost c
-                  WHERE c.account_id IS NULL
-                  AND c.tag_pool_id IS NULL
-                  AND c.message_direction = %(message_direction)s
+                  LEFT OUTER JOIN billing_tagpool t ON (c.tag_pool_id = t.id)
+                  LEFT OUTER JOIN billing_account a ON (c.account_id = a.id)
+                  WHERE
+                      (a.account_number = %(account_number)s OR
+                       c.account_id IS NULL)
+                      AND
+                      (t.name = %(tag_pool_name)s OR c.tag_pool_id IS NULL)
+                      AND
+                      (c.provider = %(provider)s OR c.provider IS NULL)
+                      AND
+                      (c.message_direction = %(message_direction)s)
             ) as t
-            ORDER BY t.account_number
+            ORDER BY
+                t.account_number NULLS LAST,
+                t.tag_pool_name NULLS LAST
```

Comment:
Interesting that the test for this passed -- maybe it was passing because the MessageCost row with null provider was created last? Ordering in SQL results is tricky to test.